packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +498,406 @@ class CommonFinders {
+      view: view,
+    );
+  }
+
+  bool _matchesPattern(String target, Pattern pattern) {
+    if (pattern is RegExp) {
+      return pattern.hasMatch(target);
+    } else {
+      return pattern == target;
+    }
+  }
+
+  SemanticsNode _rootFromView(FlutterView? view) {
+    view ??= TestWidgetsFlutterBinding.instance.platformDispatcher.implicitView;
+    final RenderView renderView = TestWidgetsFlutterBinding.instance.renderViews
+      .firstWhere((RenderView r) => r.flutterView == view);
+
+    return renderView.owner!.semanticsOwner!.rootSemanticsNode!;
+  }
+}
+
+/// Describes how a string of text should be pluralized.
+enum Plurality {
+  /// Text should be pluralized to describe zero items.
+  zero,
+  /// Text should be pluralized to describe a single item.
+  one,
+  /// Text should be pluralized to describe more than one item.
+  many,
+}
+
+extension _IntToPlurality on num {
+  Plurality toPlurality() {
+    if (this == 0) {
+      return Plurality.zero;
+    } else if (this == 1) {
+      return Plurality.one;
+    } else {
+      return Plurality.many;
+    }
+  }
+}
+
+/// Encapsulates the logic for searching a list of candidates and filtering the
+/// candidates to only those that meet the requirements defined by the finder.
+///
+/// Implementations will need to implement [allCandidates] to define the total
+/// possible search space and [findInCandidates] to define the requirements of
+/// the finder.
+///
+/// This library contains [WidgetFinder] and [SemanticsFinder] for searching
+/// Flutter's element and semantics trees respectively. If the search can be
+/// represented as a predicate, then use [MatchFinderMixin] along with the
+/// appropriate base class.
+abstract class FinderBase<CandidateType> {
+  bool _cached = false;
+
+  /// The results of the latest [find] or [tryFind] call.
+  ///
+  /// Unlike [find] and [tryFind], [found] will not re-execute the search for
+  /// this finder. Either [find] or [tryFind] must be called before accessing
+  /// [found].
+  FinderResult<CandidateType> get found {
+    assert(
+      _found != null,
+      'No results have been found yet. '
+      'Either `find` or `tryFind` must be called before accessing `found`',
+    );
+    return _found!;
+  }
+  FinderResult<CandidateType>? _found;
+
+  /// Whether or not this finder has any results in [found].
+  bool get hasFound => _found != null;
+
+  /// Describes zero, one, or more candidates that meet the requirements of a
+  /// finder.
+  ///
+  /// {@template flutter_test.finders.FinderBase.describeCandidate}
+  /// The description returned should be a brief English phrase describing a
+  /// candidate with the proper plural form. As an example for a string finder
+  /// that is looking for strings starting with "hello":
+  ///
+  /// ```dart
+  /// String describeCandidate(Plurality plurality) {
```

Comment:
That works for a case where we:

a) know the exact number, and
b) want the user to describe the number as well

This descriptor is reused for both the finder and its result, but the result is the only place where we have the number and want to use the number, hence the Plurality enum.

I'll agree that it's not my favorite solution, but the other solution is something like documenting that "when `count` is null, don't describe the amount of candidates, but otherwise include the number in the description" which is _particularly_ unintuitive when you're making a new finder imo.

Reply:
Resolving per in-person discussion.